File: ios/Classes/IMFlutterApiImpl.swift
//
//  IMFlutterApiImplementation.swift
//  halo_im
//
//  Created by Dale on 2024/9/11.
//

import Flutter
import UIKit
import XMTP
import LibXMTP

class IMFlutterApiImplementation {
    let messageApi: IMFlutterApi
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.messageApi = IMFlutterApi(binaryMessenger: binaryMessenger)
    }
    
    func flutterTest(str: String) {
        DispatchQueue.main.async {
            self.messageApi.flutterApiTest(test: str) { _ in }
        }
    }
    
    func onSignMessage(requestId requestIdArg: String, message messageArg: String) {
        DispatchQueue.main.async {
            self.messageApi.onSignMessage(requestId: requestIdArg, message: messageArg) { _ in}
        }
    }
    
    func onSubscribeToConversations(inboxIdArg: String, mArg: IMConversation) {
        DispatchQueue.main.async {
            self.messageApi.onSubscribeToConversations(inboxId: inboxIdArg, conversation: mArg) { _ in }
        }
    }
    
    func onSubscribeToMessages(inboxIdArg: String, mArg: IMDecryptedMessage) {
        DispatchQueue.main.async {
            self.messageApi.onSubscribeToMessages(inboxId: inboxIdArg, message: mArg) { _ in }
        }
    }
    
    func onSubscribeToAllMessages(inboxIdArg: String, mArg: IMDecryptedMessage) {
        DispatchQueue.main.async {
            self.messageApi.onSubscribeToAllMessages(inboxId: inboxIdArg, message: mArg) { _ in }
        }
    }
}

File: android/src/main/kotlin/halo/social/halo_im/api/IMFlutterApiImpl.kt
package halo.social.halo_im.api

import IMConversation
import IMDecryptedMessage
import IMFlutterApi
import android.util.Log
import androidx.annotation.UiThread
import io.flutter.plugin.common.BinaryMessenger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class IMFlutterApiImpl {
    companion object {
        var api: IMFlutterApi? = null
        fun setUp(binaryMessenger: BinaryMessenger) {
            Log.i("IMFlutterApiImpl", "setUp")
            if (api == null) {
                api = IMFlutterApi(binaryMessenger)
            }
        }
    }

    fun onSubscribeToConversations(inboxIdArg: String, mArg: IMConversation) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                api?.onSubscribeToConversations(inboxIdArg, mArg) {}
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun onSubscribeToMessages(inboxIdArg: String, message: IMDecryptedMessage) {
        Log.i("IMFlutterApiImpl", "onSubscribeToMessages")
        Log.i("IMFlutterApiImpl", "${message.content}")
        CoroutineScope(Dispatchers.Main).launch {
            api?.onSubscribeToMessages(inboxIdArg, message) {}
        }
    }

    fun flutterTest(str: String) {
        Log.d("IMFlutterApiImpl", "flutterTest A $str  Thread - ${Thread.currentThread().name}")
        try {
            api?.flutterApiTest(str) {}
        } catch (e: Exception) {
            e.printStackTrace()
        }
        Log.d("IMFlutterApiImpl", "flutterTest B $str  Thread - ${Thread.currentThread().name}")
    }

    fun onSubscribeToAllMessages(
        inboxIdArg: String,
        messageArg: IMDecryptedMessage,
        callback: (Result<Unit>) -> Unit
    ) {
        CoroutineScope(Dispatchers.Main).launch {
            api?.onSubscribeToAllMessages(inboxIdArg, messageArg) {}
        }
    }

    fun onSignMessage(
        id: String,
        message: String,
    ) {
        CoroutineScope(Dispatchers.Main).launch {
            api?.onSignMessage(id, message){}
        }
    }


}